In [127]:
import pandas as pd
from pathlib import Path

In [128]:
# Data path to the raw data
DATA_PATH = Path("../data/raw/Customer-Churn-Dataset.xlsx")

In [129]:
# Read the excel data

try:
    df = pd.read_excel(DATA_PATH)
    display(df.head())
except FileNotFoundError:
    print(f"Error: The file was not found at {DATA_PATH}")
except Exception as e:
    print(f"An error occurred while reading the file: {e}")

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... StreamingTV  \
0  No phone service             DSL             No  ...          No   
1                No             DSL            Yes  ...          No   
2                No             DSL            Yes  ...          No   
3  No phone service             DSL            Yes  ...          No   
4                No     Fiber optic             No  ...          No   

  StreamingMovies        Contract PaperlessBilling              PaymentMethod  \
0              No  Month-to-month              Yes           Electronic check   
1              No        One year               No               Mailed check   
2              No  Month-to-month              Yes               Mailed check   
3              No        One year               No  Bank transfer (automatic)   
4              No  Month-to-month              Yes           Electronic check   

  MonthlyCharges TotalCharges numAdminTickets  numTechTickets Churn  
0          29.85        29.85               0               0    No  
1          56.95       1889.5               0               0    No  
2          53.85       108.15               0               0   Yes  
3          42.30      1840.75               0               3    No  
4          70.70       151.65               0               0   Yes  

[5 rows x 23 columns]

In [130]:
# Display column information
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 23 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


### All the columns are in the right Dtype except the `TotalCharges`.

### `TotalCharges` should be in numeric specifically a `float` Dtype.

In [131]:
# Display dimensions of the DataFrame
print("\nShape of the DataFrame:", df.shape)


Shape of the DataFrame: (7043, 23)


In [132]:
# Check for missing values
print("\nMissing values per column:\n", df.isnull().sum())


Missing values per column:
 customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
numAdminTickets     0
numTechTickets      0
Churn               0
dtype: int64


In [134]:
# Convert 'TotalCharges' to numeric, coercing errors to NaN
df['TotalCharges'] = pd.to_numeric(df['TotalCharges'], errors='coerce')

In [ ]:
# Check for missing values
print("\nMissing values per column:\n", df.isnull().sum())


Missing values per column:
 customerID           0
gender               0
SeniorCitizen        0
Partner              0
Dependents           0
tenure               0
PhoneService         0
MultipleLines        0
InternetService      0
OnlineSecurity       0
OnlineBackup         0
DeviceProtection     0
TechSupport          0
StreamingTV          0
StreamingMovies      0
Contract             0
PaperlessBilling     0
PaymentMethod        0
MonthlyCharges       0
TotalCharges        11
numAdminTickets      0
numTechTickets       0
Churn                0
dtype: int64


### Converting `TotalCharges` to float values has introduced missing value in this feature.

In [ ]:
# Handle missing values in 'TotalCharges' - for now, we'll fill with the median as an example
# A more in-depth analysis of these rows might be needed in a real scenario
median_total_charges = df['TotalCharges'].median()
df['TotalCharges'] = df['TotalCharges'].fillna(median_total_charges)

In [ ]:
df['TotalCharges'].isna().sum()

np.int64(0)

In [ ]:
# Verify that there are no more missing values in TotalCharges
print("\nMissing values after filling NaN in TotalCharges:\n", df.isnull().sum())


Missing values after filling NaN in TotalCharges:
 customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
numAdminTickets     0
numTechTickets      0
Churn               0
dtype: int64


### Save the clean Dataset

In [ ]:
# define clean data path 

clean_data_path = Path("../data/processed/Clean_Customer-Churn-Dataset.csv")
df.to_csv(clean_data_path, index=False)

## Data Exploration Summary

Based on the initial data inspection and cleaning:

*   The dataset contains information on 7043 customers with 23 columns.
*   We checked for missing values and found that the `TotalCharges` column initially had no missing values.
*   The `TotalCharges` column was converted from an object data type to a numeric type.
*   The missing values in the `TotalCharges` column were imputed with the median value of the column.
*   There are no other missing values in the dataset.
